---
title: "Top GEX Kallisto mef Mouse"
output: html_notebook
editor_options: 
  chunk_output_type: inline
---

```{r setup}
suppressPackageStartupMessages({if (!require("BiocManager", quietly = TRUE))
    install.packtimepoints("BiocManager")
})

pacman::p_load(dplyr, data.table, ggplot2, tidyverse, scater, BiocManager, readr, DESeq2, pheatmap, RColorBrewer, tximport, GenomicFeatures, org.Mm.eg.db, AnnotationDbi, ggrepel)
```


#PREPARING DATA
```{r}
# Read in salmon abundance files

dir <- paste0(getwd(), "/kallisto_quant_mouse_mef")

list.files(dir)
```



```{r}
files <- file.path(dir, c("SRR28878885.tabular", 
                          "SRR28878886.tabular",
                          "SRR28878887.tabular"))
                   
all(file.exists(files))

names(files) <- c("mef_tpm_1", "mef_tpm_2", "mef_tpm_3")

files
```


#tx2gene

```{r}
TxDb <- makeTxDbFromGFF(file = "C:/Users/gevd910d/Desktop/Hypoxia MS Final/A1 Analysis/A1_C2C12_myoblast_comparison/gencode.vM35.annotation.gtf.gz")
                        
k <- keys(TxDb, keytype = "TXNAME")

tx2gene <- AnnotationDbi::select(TxDb, k, "GENEID", "TXNAME", "")
```



```{r}
txi.kallisto <- tximport(files, 
                         type = "kallisto",
                         tx2gene = tx2gene,
                         ignoreAfterBar = TRUE) 

abundance <- txi.kallisto$abundance # Extract TPM
```




# ANNOTATION
```{r}
abundance <- as.data.frame(abundance) %>%
  rownames_to_column(var = "GENEID")  %>%
  mutate(GENEID = sub("\\..*", "", GENEID)) %>%
  column_to_rownames(var = "GENEID") %>%
  as.matrix()




median_tpm <-  rowMedians(abundance) 

median_tpm <- data.frame(count = median_tpm, 
                         GENEID = rownames(abundance))
```


```{r}
annotations <- AnnotationDbi::select(org.Mm.eg.db, 
                      keys = rownames(abundance), 
                      columns = c("ENTREZID", "GENENAME", "SYMBOL"),
                      keytype = "ENSEMBL",
                      multiVals = "first")

median_tpm <- merge(annotations, median_tpm, by.x="ENSEMBL", by.y="GENEID")
```





```{r}
median_tpm$count <- round(median_tpm$count, 2)


anno_tpm <- median_tpm %>% mutate(logmedian = log2(count + 1))

anno_tpm$logmedian <- round(anno_tpm$logmedian, 2)
```




```{r}
marker_tpm <- anno_tpm %>%
  mutate(count = round(count)) %>%
  dplyr::select(ENSEMBL, SYMBOL, count, logmedian) %>%
  dplyr::filter(SYMBOL %in% 
                  
                  c("Tagln", "Acta2", "Cald1", "Des", "Aebp1", "Myocd", "Myh11", "Myh9", "Myh9", "Ccn1", "Palld", "Myl9", "Cnn1", "Tpm2", "Tpm1", "Tagln2", "Mcam", "Tinagl1", "Flna", "Myog", "Myf6",
                    
                    "Col5a2", "Loxl1", "Pdgfra", "Fbln1", "Lum", "Dpt", "Hsd11b1", "CD34", "GSN", "Fbln1", "Fbln2",
                    
                    
                    "Col1a2", "Col1a1", "Timp2", "Col12a1", "Col6a2", "Vim", "Fn1", "Col5a1", "Rgs4", "Mgst3", "Ccn2",
                    
                    "Pax3", "Pax7", "Myod1", "Myf5"
                    
                    
                    )) %>%
  
  
  dplyr::mutate(cell_type = dplyr::case_when(
    SYMBOL %in% c("Tagln", "Acta2", "Cald1", "Des", "Aebp1", "Myocd", "Myh11", "Myh9", "Myh9", "Ccn1", "Palld", "Myl9", "Cnn1", "Tpm2", "Tpm1", "Tagln2", "Mcam", "Tinagl1", "Flna", "Myog", "Myf6") ~ "Smooth Muscle",
    
    SYMBOL %in% c("Col5a2", "Loxl1", "Pdgfra", "Fbln1", "Lum", "Dpt", "Hsd11b1", "CD34", "GSN", "Fbln1", "Fbln2") ~ "Fibroblast",
    
    SYMBOL %in% c("Col1a2", "Col1a1", "Timp2", "Col12a1", "Col6a2", "Vim", "Fn1", "Col5a1", "Rgs4", "Mgst3", "Ccn2") ~ "Mixed",
    
    SYMBOL %in% c("Pax3", "Pax7", "Myod1", "Myf5") ~ "Myoblast"

  )) %>%
  
  dplyr::arrange(cell_type, desc(count)) %>%
  mutate(normalized_tpm = round(100 * (logmedian - min(logmedian)) / (max(logmedian) - min(logmedian))))
```


```{r}
# Set the order of cell_type
marker_tpm$cell_type <- factor(marker_tpm$cell_type, 
                               levels = c("Smooth Muscle", "Fibroblast", "Mixed", "Myoblast"))

marker_tpm <- marker_tpm %>%
  arrange(cell_type, desc(logmedian)) %>%
  mutate(SYMBOL = factor(SYMBOL, levels = unique(SYMBOL)))

# Define colors
color_palette <- c("Smooth Muscle" = "#F8766D", 
                   "Fibroblast" = "#00BFC4", 
                   "Mixed" = "#7CAE00", 
                   "Myoblast" = "#C77CFF")
```



```{r}
# Log2 scale

ggplot(marker_tpm, aes(x = SYMBOL, y = logmedian, fill = cell_type)) +
  geom_bar(stat = "identity") + 
  
    geom_text_repel(aes(label = SYMBOL), 
                  size = 5, 
                  vjust = -0.2,  
                  hjust = 0.0,   
                  angle = 40,    
                  box.padding = 0.3,  # Adjust padding between labels
                  point.padding = 0.2,  # Adjust padding between points and labels
                  segment.size = 0,  # Remove the lines between label and bar
                  show.legend = FALSE) + 
  
  geom_hline(yintercept = 3.32, linetype = "dashed", color = "black") + # 10 TPM
  geom_hline(yintercept = 9.97, linetype = "dashed", color = "black") + # 1000 TPM
  
  annotate("text", x = Inf, y = 9.97, label = "1000 TPM", vjust = -0.5, hjust = 1.1, color = "red", size = 3.5) + # Add label to the dashed line
  
  annotate("text", x = Inf, y = 3.32, label = "10 TPM", vjust = -0.5, hjust = -0.5, color = "red", size = 3.5) + # Add label to the dashed line
  
  scale_fill_manual(values = color_palette) +  # Apply the color palette
  
  theme_minimal() +
  labs(title = "Log2 Median TPM Across 3 MEF Samples",
       x = "Gene Symbol",
       y = "Log2 Median TPM") +
  ylim(0, NA) +
   scale_y_continuous(breaks = seq(0, ceiling(max(marker_tpm$logmedian)), by = 3)) +  # Add ticks every 3 units
  
  coord_cartesian(clip = "off") +  # Allow plotting outside the panel
  
  theme(legend.title = element_blank(),
        legend.position = "none",
        axis.text.x = element_blank(),  # Remove x-axis labels
        axis.ticks.x = element_blank(),
        plot.title = element_text(size = 20),  # Increase title size
        axis.title.x = element_text(size = 16),  # Increase x-axis label size
        #axis.title.y = element_blank(),
        
        legend.text = element_text(size = 12),  # Increase legend text size
        legend.key.size = unit(1, "cm"),      # Make legend keys bigger
        legend.spacing.y = unit(0.5, "cm"))     # Add more spacing between legend items

ggsave(filename = "Log2 Median TPM 3 MEF Samples.png", 
       bg = "transparent",
       width = 10, height = 12, dpi = 1200)
```





#END

```{r}
sessionInfo()
```

